ironfish-rust/src/serializing/mod.rs

```diff
@@ -0,0 +1,101 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+pub mod aead;
+
+use crate::errors::IronfishError;
+
+/// Helper functions to convert pairing parts to bytes
+///
+/// The traits in the pairing and zcash_primitives libraries
+/// all have functions for serializing, but their interface
+/// can be a bit clunky if you're just working with bytearrays.
+use ff::PrimeField;
+use group::GroupEncoding;
+
+use std::io;
+
+const HEX_CHARS: &[u8; 16] = b"0123456789abcdef";
+
+pub(crate) fn read_scalar<F: PrimeField, R: io::Read>(mut reader: R) -> Result<F, IronfishError> {
+    let mut fr_repr = F::Repr::default();
+    reader.read_exact(fr_repr.as_mut())?;
+
+    Option::from(F::from_repr(fr_repr)).ok_or(IronfishError::InvalidData)
+}
+
+pub(crate) fn read_point<G: GroupEncoding, R: io::Read>(mut reader: R) -> Result<G, IronfishError> {
+    let mut point_repr = G::Repr::default();
+    reader.read_exact(point_repr.as_mut())?;
+
+    Option::from(G::from_bytes(&point_repr)).ok_or(IronfishError::InvalidData)
+}
+
+/// Output the bytes as a hexadecimal String
+pub fn bytes_to_hex(bytes: &[u8]) -> String {
+    let mut hex: Vec<u8> = vec![0; bytes.len() * 2];
+
+    for (i, b) in bytes.iter().enumerate() {
+        hex[i * 2] = HEX_CHARS[(b >> 4) as usize];
+        hex[i * 2 + 1] = HEX_CHARS[(b & 0x0f) as usize];
+    }
+
+    unsafe { String::from_utf8_unchecked(hex) }
+}
+
+/// Output the hexadecimal String as bytes
+pub fn hex_to_bytes<const SIZE: usize>(hex: &str) -> Result<[u8; SIZE], IronfishError> {
+    if hex.len() != SIZE * 2 {
+        return Err(IronfishError::InvalidData);
+    }
+
+    let mut bytes = [0; SIZE];
+
+    let hex_iter = hex.as_bytes().chunks_exact(2);
+
+    for (i, hex) in hex_iter.enumerate() {
+        bytes[i] = hex_to_u8(hex[0])? << 4 | hex_to_u8(hex[1])?;
+    }
+
+    Ok(bytes)
+}
```

Comment:
`hex_to_bytes` now takes an array, relying on a const generic to determine the size. Removes any heap allocations via `Vec`s.

At first glance, this looks like it would be more painful to use, however, we were ALWAYS using it with an array already, but had to do a lot of extra work moving back and forth between arrays and vecs. You'll notice the call-sites are all a LOT cleaner.